domain/estimators/recurrent_logistic.py

```diff
@@ -0,0 +1,378 @@
+from abc import ABCMeta, abstractmethod
+import copy
+import logging
+
+import pandas as pd
+import time
+import torch
+from torch.utils.data import TensorDataset, DataLoader
+from tqdm import tqdm
+
+from ..estimator import Estimator
+from utils import dictify_df
+
+
+class RecurrentLogistic(Estimator, torch.nn.Module):
+    """
+    RecurrentLogistic is an Estimator that approximates posterior of
+    p(v_cur | v_init) by training a logistic regression model to predict the current
+    value in a cell given all other initial values using features
+    of the other initial values such as co-occurrence.
+    """
+
+    def __init__(self, dataset, pruned_domain, active_attrs):
+        """
+        :param dataset: (Dataset) original dataset
+        :param pruned_domain: (dict) of tid -> attr -> value for the domain
+        :param active_attrs: (list[str]) attributes that have random values
+        """
+        torch.nn.Module.__init__(self)
+        Estimator.__init__(self, dataset)
+
+        self.dom = pruned_domain
+        self.active_attrs = active_attrs
+
+        # self.dom maps tid --> attr --> list of domain values
+        # we need to find the number of domain values we will be generating
+        # a training sample for.
+        self.n_samples = sum(len(dom) for _, attrs in self.dom.items() for _, dom in attrs.items())
+
+        # Make a copy of the raw data as our current data.
+        self.cur_df = self.ds.get_raw_data().copy()
+
+        # Generate featurizers for this model.
+        self._update_featurizers(reuse_stats=True)
+
+        # Make a copy of the initial featurizers for prediction/test set
+        # i.e., we want to use our original co-occurrence statistics.
+        self.test_featurizers = [feat.copy() for feat in self.train_featurizers]
+
+        # Use pytorch logistic regression model.
+        self._W = torch.nn.Parameter(torch.zeros(self.num_features, 1))
+        torch.nn.init.xavier_uniform_(self._W)
+        self._B = torch.nn.Parameter(torch.Tensor([1e-6]))
+        self._loss = torch.nn.BCELoss()
+        self._optimizer = torch.optim.Adam(self.parameters())
+
+    def _update_featurizers(self, reuse_stats=False):
+        """
+        Reinitialize featurizers that depend on updated current values.
+
+        :param reuse_stats: (bool) reuses frequency and co-occurrence statistics from
+            raw Dataset instead of re-computing.
+        """
+
+        # Featurizers used for training.
+        freq, cooccur_freq = None, None
+        if reuse_stats:
```

Comment:
Has the data changed in the raw dataset. Why might we want to recompute stats here?